subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java

```diff
@@ -22,74 +22,215 @@
-                        return Optional.empty();
+                    public ExecutionOutcome getOutcome() {
+                        return skipOutcome;
                     }
 
                     @Override
-                    public Duration getDuration() {
-                        return Duration.ZERO;
+                    public Object getOutput() {
+                        return work.loadRestoredOutput(context.getWorkspace());
                     }
-                };
-            })
-            .orElseGet(() -> delegate.execute(work, context));
+                });
+            }
+
+            @Override
+            public CachingState getCachingState() {
+                return CachingState.NOT_DETERMINED;
+            }
+
+            @Override
+            public ImmutableList<String> getExecutionReasons() {
+                return ImmutableList.of();
+            }
+
+            @Override
+            public Optional<AfterExecutionState> getAfterExecutionState() {
+                return Optional.empty();
+            }
+
+            @Override
+            public Optional<OriginMetadata> getReusedOutputOriginMetadata() {
+                return Optional.empty();
+            }
+        };
+    }
+
+    private CachingResult executeWithNoEmptySources(UnitOfWork work, PreviousExecutionContext context, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> newInputFileProperties) {
+        return executeWithNoEmptySources(work, new PreviousExecutionContext() {
+            @Override
+            public ImmutableSortedMap<String, CurrentFileCollectionFingerprint> getInputFileProperties() {
+                return newInputFileProperties;
+            }
+
+            @Override
+            public Optional<PreviousExecutionState> getPreviousExecutionState() {
+                return context.getPreviousExecutionState();
+            }
+
+            @Override
+            public File getWorkspace() {
+                return context.getWorkspace();
+            }
+
+            @Override
+            public Optional<ExecutionHistoryStore> getHistory() {
+                return context.getHistory();
+            }
+
+            @Override
+            public ImmutableSortedMap<String, ValueSnapshot> getInputProperties() {
+                return context.getInputProperties();
+            }
+
+            @Override
+            public UnitOfWork.Identity getIdentity() {
+                return context.getIdentity();
+            }
+
+            @Override
+            public Optional<String> getNonIncrementalReason() {
+                return context.getNonIncrementalReason();
+            }
+
+            @Override
+            public WorkValidationContext getValidationContext() {
+                return context.getValidationContext();
+            }
+        });
+    }
+
+    private CachingResult executeWithNoEmptySources(UnitOfWork work, PreviousExecutionContext context) {
```

Comment:
"with no empty sources" is a bit weird, I think `executeWithSourcesPresent()` or even `executeWithNonEmptySources()` or even even `executeWithoutEmptySources()` would all be better.